src/i2c.rs

```diff
@@ -519,21 +572,63 @@ where
     }
 }
 
-struct CommandLink(i2c_cmd_handle_t);
+#[repr(u32)]
+enum AckType {
+    Ack = i2c_ack_type_t_I2C_MASTER_ACK,
+    #[allow(dead_code)]
+    Nack = i2c_ack_type_t_I2C_MASTER_NACK,
+    LastNack = i2c_ack_type_t_I2C_MASTER_LAST_NACK,
+}
+
+struct CommandLink<'buffers>(i2c_cmd_handle_t, PhantomData<&'buffers u8>);
 
-impl CommandLink {
+impl<'buffers> CommandLink<'buffers> {
```

Comment:
I don't think the `'buffers` lifetime is used for anything. Also given that `CommandLink` is a private structure, why bother? Perhaps you could remove it?

Reply:
The `'buffers` lifetime is used to make sure the references given in `master_read` and `master_write` live as long as the `CommandLink`, since that's the contract for the underlying `i2c_master_*` functions.
The alternative would be to remove the lifetime annotation and mark those methods as unsafe. Which is not quite as nice since you now have unsafe everywhere you call those methods.

Simply removing the lifetime and ignoring it would defeat the point of using Rust, no? It's good to have the compiler verify that what we're doing internally is correct.

Also, the annotation will be useful for the `transaction_iter` variant where ownership is more of a concern I think.

Even though it's a private type, I think leaving it in does little harm. The compiler does well to infer the lifetime parameter at the call site.

Reply:
> The `'buffers` lifetime is used to make sure the references given in `master_read` and `master_write` live as long as the `CommandLink`, since that's the contract for the underlying `i2c_master_*` functions. The alternative would be to remove the lifetime annotation and mark those methods as unsafe. Which is not quite as nice since you now have unsafe everywhere you call those methods.
> 
> Simply removing the lifetime and ignoring it would defeat the point of using Rust, no? It's good to have the compiler verify that what we're doing internally is correct.

Not quite. I think you are unnecessarily complicating it. If you remove the `&'buffers` annotation from `master_read` and `master_write`, you'll switch into ellided lifetimes, where the lifetime of the buffer will be equal to the lifetime of the `&mut self` instance, which is shorter in duration than `'buffers` and just as good for that purpose. And simpler.

> The alternative would be to remove the lifetime annotation and mark those methods as unsafe. Which is not quite as nice since you now have unsafe everywhere you call those methods.

That's not true. Remove the `'buffers` annotation, and you'll see that the compiler is not forcing to mark anything with `unsafe` that you haven't marked already.

Reply:
> where the lifetime of the buffer will be equal to the lifetime of the `&mut self` instance

`&mut self` only lives as long as the function call though. The buffer passed in needs to lives as long as `CommandLink
`. if this is to reflect the contract of the C function.

> That's not true. Remove the `'buffers` annotation, and you'll see that the compiler is not forcing to mark anything with `unsafe` that you haven't marked already.

You're right, the compiler won't force me to use `unsafe`. I should've been clearer there. Adding `unsafe` there would be for correctness (since the compiler doesn't know that it should force me), as the caller needs to know that the lifetime of the buffer should be managed with care.

Well, I can see that the real issue here, is that `CommandLink` is private so why try to make it match the C contract when we know what we're doing in the HAL. Would you be alright with the `unsafe` alternative?

Reply:
> > where the lifetime of the buffer will be equal to the lifetime of the `&mut self` instance
> 
> `&mut self` only lives as long as the function call though. The buffer passed in needs to lives as long as `CommandLink `. if this is to reflect the contract of the C function.

Well... I re-read the ESP-IDf docu for `i2c_master_read` and `i2c_master_write` and you are right! The buffers **must** remain valid even after the call to `i2c_master_*` is over.

So I'll merge the MR as-is.